Update focal cn file preparation (#95)
PediatricOpenTargets/OpenPBTA-analysis: analyses/focal-cn-file-preparation/01-add-ploidy-cnvkit.Rmd

### Read in data

```{r}

> cnvkit_file <- file.path("..", "..", "data", "v8", "cnv-cnvkit.seg.gz")

Reviewer: ```suggestion
cnvkit_file <- file.path("..", "..", "data",  "cnv-cnvkit.seg.gz")
```
We don't need the v8 here so that it can be generic and don't need to update with every release.

Author: Done.

cnvkit_df <- readr::read_tsv(cnvkit_file)
```

```{r}

> histologies_file <- file.path("..", "..", "data", "v8", "histologies.tsv")

Reviewer: ```suggestion
histologies_file <- file.path("..", "..", "data", "histologies.tsv")
```
Same as comment above.

Author: Done

histologies_df <- readr::read_tsv(histologies_file, guess_max = 100000)
```

# Edits for https://github.com/PediatricOpenTargets/ticket-tracker/issues/177
# For PBTA, histologies file has germline_sex_estimate already assigned to WXS patients which have had WGS performed
# Updating here germline_sex_estimate for cohort=="TARGET" where is.na(germline_sex_estimate)

> #Isolate IDs where germline_sex_estimate is NA and cohort == "TARGET"

Reviewer 2: see my comment below about the cohort selection

Author: The code now uses sample_type in combination with experimental_strategy, instead of specific cohort selection.

> germline_sex_estimate_na <- histologies_df %>% filter(cohort == "TARGET") %>% filter(is.na(germline_sex_estimate))

Reviewer: ```suggestion
```{r}
germline_sex_estimate_na <- histologies_df %>% filter(cohort == "TARGET") %>% filter(is.na(germline_sex_estimate))
```

Reviewer: Actually as I was going through this more, I think here we need to do 
```
histologies_df %>% filter(experimental_strategy == "WXS") %>% 
filter(sample_type == "Tumor) %>%
filter(is.na(germline_sex_estimate))
```
I didn't add the cohort since we can do a cohort-less workflow per @jharenza's suggestion.

Author: I reviewed the code and implemented all the appropriate changes, and verified that it creates ```cnvkit_with_status.tsv``` in the ```scratch``` folder with expected values for ```germline_sex_estimate```.

> germline_sex_estimate_na <- histologies_df %>% filter(cohort == "TARGET") %>% filter(is.na(germline_sex_estimate))
> na_ids <- germline_sex_estimate_na %>% pull(Kids_First_Biospecimen_ID)
> 
> #Split the dataset into two groups, fix data for select IDs
> histology_subset1 <- histologies_df %>% filter(!Kids_First_Biospecimen_ID %in% na_ids)
> 
> histology_subset2 <- germline_sex_estimate_na %>% mutate(germline_sex_estimate = reported_gender)

Reviewer 2: ```suggestion
germline_sex_estimate_na <- histologies_df %>% 
filter(cohort == "TARGET" & is.na(germline_sex_estimate))

na_ids <- germline_sex_estimate_na %>% 
pull(Kids_First_Biospecimen_ID)

#Split the dataset into two groups, fix data for select IDs
histology_with_sex <- histologies_df %>% 
filter(!Kids_First_Biospecimen_ID %in% na_ids)

histology_no_sex <- germline_sex_estimate_na %>% 
mutate(germline_sex_estimate = reported_gender)
```

hi @sangeetashukla (and cc @runjin326 because I noticed this on some of your PRs as well) - I suggest we do not use a long row format for commands, but split these by row for readability  - this also helps with reviewing lines. So, I wouldn't commit this from the web, but rather, update the code in R so you can get the styling/indentation correct. I also got rid of one of your redundant filters. Also tried to make the df naming more descriptive so it is simple for others to come in and pick up later.

I also have a suggestion for this in general - I think we should update this without regard for cohort. For instance, we may get other cohorts of WXS (in fact, we will) and instead of adding cohorts iteratively, I suggest we have some different logic here. The logic would be if `experimental_strategy == WXS & is.na(germline_sex_estimate)`, then does WGS for this same sample exist (via `sample_id`) and does WGS `germline_sex_estimate` exist? if yes, mutate to use WGS `germline_sex_estimate`. If no, mutate to use `reported_gender.`

Reviewer: @jharenza, thanks for the comment! I will break my lines from now on!
And to you comment, yes, it makes sense to check WGS - but as I mentioned in my previous comment & confirmed by Krutika, looks like PBTA already have WGS fixed. 
And @sangeetashukla and I also checked that for TARGET, no samples with `germline_sex_estimate == NA` has matching WGS. I just did a full check for all cohorts - looks like none of them have matching WGS in PBTA and TRAGET? But maybe we can pull from GMKF? 
```
> na_sex_participant <- histology %>% filter(experimental_strategy == "WXS") %>% filter(sample_type == "Tumor") %>% filter(is.na(germline_sex_estimate))  %>% pull(Kids_First_Participant_ID)
> histology %>% filter(Kids_First_Participant_ID %in% na_sex_participant) %>% select(cohort, experimental_strategy) %>% table()
        experimental_strategy
cohort   RNA-Seq  WGS  WXS
  GMKF         8   20    0
  PBTA        18    0   32
  TARGET     531    0 1286
```
We can definitely add the check for WGS in the workflow and change to gender when WGS does not exist still.

Reviewer 2: >And to you comment, yes, it makes sense to check WGS - but as I mentioned in my previous comment & confirmed by Krutika, looks like PBTA already have WGS fixed.

Oh yes, PBTA is currently OK, but in v10, we will have some additional WXS from PBTA which will not have WGS, so would be good to implement this logic now.

Reviewer: @jharenza, good point! Let's implemented the WGS logic then! (cc @sangeetashukla)

Author: Done

#Merge data after fixing germline_estimate with new value

> histologies_df_final <- rbind(histology_subset1,histology_subset2)

Reviewer 2: ```suggestion
histologies_df_final <- bind_rows(histology_with_sex, histology_no_sex)
```

rbind is deprecated, so you should switch to `bind_rows` here; also added spacing

Author: Implemented the change.

> 

Reviewer: In here we need to add "```" to close the chunk - this is a R markdown file, and if we did not put them into these brackets - they would not run and will throw an error when we run the code as followed:
```
Error in select(., Kids_First_Biospecimen_ID, tumor_ploidy, germline_sex_estimate) : 
  object 'histologies_df_final' not found
```

Author: Thank you.

#Verify counts match
#length(histologies_df$Kids_First_Biospecimen_ID)
#length(histology_subset1$Kids_First_Biospecimen_ID)
#length(histology_subset2$Kids_First_Biospecimen_ID)

> #length(histologies_df_final$Kids_First_Biospecimen_ID)

Reviewer 2: This should be removed since it is commented out. Since you are using a notebook here, you can add a true/false logic statement to determine whether the counts match and print this out.

Author: I removed those lines. The code ensures the histologies data is complete by implementing appropriate joins.

histologies_df <- readr::read_tsv(histologies_file, guess_max = 100000)
```

# Edits for https://github.com/PediatricOpenTargets/ticket-tracker/issues/177
```{r}
# Filter to WXS tumor with no germline sex estimate

> no_estimate <- histologies_df %>% filter(sample_type == "Tumor") %>% filter(experimental_strategy == "WXS") %>%   filter(is.na(germline_sex_estimate)) %>% select(-germline_sex_estimate)

Reviewer: ```suggestion
no_estimate <- histologies_df %>% 
filter(sample_type == "Tumor") %>% 
filter(experimental_strategy == "WXS") %>%  
 filter(is.na(germline_sex_estimate)) %>%
  select(-germline_sex_estimate)
```

# Extract participant ID with no germline sex estimate

> no_estimate_participants <- no_estimate %>% pull(cohort_participant_id) %>% unique()

Reviewer: ```suggestion
no_estimate_participants <- no_estimate %>% 
pull(cohort_participant_id) %>% 
unique()
```

# Find out matching WGS - currently this has no entry for v9

> no_estimate_participants_match_wgs <- histologies_df %>% filter(cohort_participant_id %in% no_estimate_participants) %>% filter(experimental_strategy == "WGS") %>% filter(sample_type == "Tumor") %>% select(cohort_participant_id, germline_sex_estimate)

Reviewer: ```suggestion
no_estimate_participants_match_wgs <- histologies_df %>% 
filter(cohort_participant_id %in% 
no_estimate_participants) %>% 
filter(experimental_strategy == "WGS") %>% 
filter(sample_type == "Tumor") %>% 
select(cohort_participant_id, germline_sex_estimate)
```

# Merge data that was fixed with WGS germline sex estimate

> no_estimate_fixed_through_wgs <- left_join(no_estimate, no_estimate_participants_match_wgs, by = "cohort_participant_id") %>% filter(!is.na(germline_sex_estimate))

Reviewer: ```suggestion
no_estimate_fixed_through_wgs <- left_join(no_estimate, no_estimate_participants_match_wgs, by = "cohort_participant_id") %>% 
filter(!is.na(germline_sex_estimate))
```

# Identify data that was not fixed due to NA
no_estimate_still_na <- left_join(no_estimate, no_estimate_participants_match_wgs, by = "cohort_participant_id") %>% filter(is.na(germline_sex_estimate))

# Fix NA data using gender
no_estimate_still_na_fixed <- no_estimate_still_na %>% 
  mutate(germline_sex_estimate = case_when(
                                  reported_gender == "Female" ~ "Female", 
                                  reported_gender == "Male" ~ "Male",
                                  TRUE ~ as.character(NA))
         )

# Merge back fixed_wgs, fixed_na, and exist_germline_estimate data
histologies_df_fixed <- bind_rows(no_estimate_fixed_through_wgs, no_estimate_still_na_fixed)

histologies_df_final <- histologies_df %>% 
  left_join(select(histologies_df_fixed, germline_sex_estimate, Kids_First_Biospecimen_ID), by = "Kids_First_Biospecimen_ID",
            suffix = c('.x', '.y'), suffixes_mandatory = c(FALSE, TRUE)) %>%

>   mutate(germline_sex_estimate = case_when(Kids_First_Biospecimen_ID == Kids_First_Biospecimen_ID ~ germline_sex_estimate.y

Reviewer: Could you please also fix the indent here for readability?

Author: @runjin326  Done. Please review when you get a chance.

histologies_df <- readr::read_tsv(histologies_file, guess_max = 100000)
```

# Edits for https://github.com/PediatricOpenTargets/ticket-tracker/issues/177
```{r}
# Filter to WXS tumor with no germline sex estimate
no_estimate <- histologies_df %>% 
  filter(sample_type == "Tumor" & experimental_strategy == "WXS" &
           is.na(germline_sex_estimate)) %>%
  select(-germline_sex_estimate)

# Extract participant ID with no germline sex estimate
no_estimate_participants <- no_estimate %>% 
  pull(Kids_First_Participant_ID) %>% 
  unique()

# Find out matching WGS

> # Use cohort_participant_id for filtering samples that have WGS, when WXS was performed

Reviewer 2: ```suggestion
# Use Kids_First_Participant_ID for filtering samples that have WGS, when WXS was performed
```

no_estimate_participants_match_wgs <- histologies_df %>% 
  filter(cohort_participant_id %in% no_estimate_participants & 
           experimental_strategy == "WGS" & sample_type == "Tumor") %>% 

>   select(Kids_First_Participant_ID, germline_sex_estimate)

Reviewer 2: this needs to be changed to `Kids_First_Participant_ID`

Author: > this needs to be changed to `Kids_First_Participant_ID`

@jharenza I changed it `cohort_participant_id` to `Kids_First_Participant_ID`, but the count for GSE values is still the same. Let me know when you are reviewing this, may be we can jump on a quick call?

histologies_df <- readr::read_tsv(histologies_file, guess_max = 100000)
```

# Edits for https://github.com/PediatricOpenTargets/ticket-tracker/issues/177
```{r}
# Filter to WXS tumor with no germline sex estimate

> no_estimate <- histologies_df %>% 
>   filter(sample_type == "Tumor") %>% 
>   filter(experimental_strategy == "WXS") %>%  
>   filter(is.na(germline_sex_estimate)) %>%
>   select(-germline_sex_estimate)

Reviewer 2: ```suggestion
no_estimate <- histologies_df %>% 
  filter(sample_type == "Tumor" & experimental_strategy == "WXS" & is.na(germline_sex_estimate)) %>%
  select(-germline_sex_estimate)
```

Author: Done

# Extract participant ID with no germline sex estimate

> no_estimate_participants <- no_estimate %>% 
>   pull(cohort_participant_id) %>% 
>   unique()

Reviewer 2: ```suggestion
no_estimate_participants <- no_estimate %>% 
  pull(Kids_First_Participant_ID) %>% 
  unique()
```

I would use the `Kids_First_Participant_ID` here because that is our main ID from which we are working - theoretically these are 1:1 with `cohort_participant_id`, but the `cohort_participant_id` is metadata that "comes along" with the samples and we like to stick to KF ids.

Author: Done

> # Find out matching WGS - currently this has no entry for v9

Reviewer 2: ```suggestion
# Find out matching WGS
```

> no_estimate_participants_match_wgs <- histologies_df %>% 
>   filter(cohort_participant_id %in% 
>   no_estimate_participants) %>% 
>   filter(experimental_strategy == "WGS") %>% 
>   filter(sample_type == "Tumor") %>% 
>   select(cohort_participant_id, germline_sex_estimate)

Reviewer 2: ```suggestion
no_estimate_participants_match_wgs <- histologies_df %>% 
  filter(Kids_First_Participant_ID %in% no_estimate_participants & experimental_strategy == "WGS" & sample_type == "Tumor") %>% 
  select(Kids_First_Participant_ID, germline_sex_estimate)
```

Condensing to use `filter()` only once and add `Kids_First_Participant_ID` here. This actually gives 10 participants that match now.

Author: Done

# Merge data that was fixed with WGS germline sex estimate

> no_estimate_fixed_through_wgs <- left_join(no_estimate, no_estimate_participants_match_wgs, by = "cohort_participant_id") %>% 
>   filter(!is.na(germline_sex_estimate))
> 
> # Identify data that was not fixed due to NA
> no_estimate_still_na <- left_join(no_estimate, no_estimate_participants_match_wgs
>                                   , by = "cohort_participant_id") %>%
>   filter(is.na(germline_sex_estimate))

Reviewer 2: ```suggestion
no_estimate_fixed_through_wgs <- left_join(no_estimate, no_estimate_participants_match_wgs, by = "Kids_First_Participant_ID") %>% 
  filter(!is.na(germline_sex_estimate))

# Identify data that was not fixed due to NA
no_estimate_still_na <- left_join(no_estimate, no_estimate_participants_match_wgs
                                  , by = "Kids_First_Participant_ID") %>%
  filter(is.na(germline_sex_estimate))
```
update to use `Kids_First_Participant_ID`

# Fix NA data using gender
no_estimate_still_na_fixed <- no_estimate_still_na %>% 
  mutate(germline_sex_estimate = case_when(
                                  reported_gender == "Female" ~ "Female", 
                                  reported_gender == "Male" ~ "Male",
                                  TRUE ~ as.character(NA))
         )

# Merge back fixed_wgs, fixed_na, and exist_germline_estimate data

> histologies_df_fixed <- bind_rows(no_estimate_fixed_through_wgs, no_estimate_still_na_fixed)

Reviewer 2: Here, I am getting: 

```
> table(histologies_df_fixed$germline_sex_estimate)

Female   Male 
   155    216 
```

> histologies_df_final <- histologies_df %>% 

Reviewer 2: Here, I am getting the same n as fixed, so this merge doesn't seem to be working properly

```
> table(histologies_df_final$germline_sex_estimate)

Female   Male 
   155    216 
> table(histologies_df$germline_sex_estimate)

 Female   known    Male Unknown 
   1629      16    1892       1
```
I think this should be Female 1629+155 and Male 1892+216 - cc @runjin326

Reviewer 2: I think what you want to do here after:
```
# Merge back fixed_wgs, fixed_na, and exist_germline_estimate data
histologies_df_fixed <- bind_rows(no_estimate_fixed_through_wgs, no_estimate_still_na_fixed)
```

 is this:

```
# remove samples in histologies_df_fixed from histologies_df
histologies_df_rm <- histologies_df %>%
  anti_join(histologies_df_fixed, by = "Kids_First_Biospecimen_ID")

# add new rows back
histologies_df_final <- histologies_df_rm %>% 
  bind_rows(histologies_df_fixed)
```

Now I get the expected N of germline sex estimate:

```
> table(histologies_df_final$germline_sex_estimate)

 Female   known    Male Unknown 
   1784      16    2108       1 
> 1629+155
[1] 1784
> 1892+216
[1] 2108
```

Reviewer 2: @sangeetashukla and @runjin326 added a comment about this [here](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/95/files#r702130039) cc @afarrel

Author: @jharenza @runjin326 
It makes sense to use KF participant ID as you pointed out.
I am now using cohort_participant_id only to identify samples that have WGS performed when WXS is also performed.

However, while the split between the histologies_df and histologies_df_rm for KF Biospecimen ID looks correct; the split between male/female values for GSE in histologies_df and histologies_df_rm is the same. That is confusing to me. Any thoughts?

Reviewer 2: >I am now using cohort_participant_id only to identify samples that have WGS performed when WXS is also performed.

We should not use `cohort_participant_id` at all - can you update this please?

# this logic is consistent with what is observed in the controlfreec file
# specifically, in samples where germline sex estimate = Female, X chromosomes
# appear to be treated the same way as autosomes

samples_WGS <- histologies_df %>% filter(experimental_strategy == "WGS")
samples_WXS <- histologies_df %>% filter(experimental_strategy == "WXS")
samples_superset <- inner_join(samples_WGS, samples_WXS, by = "cohort_participant_id", copy = FALSE)

> 

Reviewer: @sangeetashukla, thanks so much for preparing for this PR. I haven't looked into why the case_when was not working but I actually envisioned this being done a bit different. I would probably first generate a new histology file just for the add_ploidy_df first as followed:
First find the WXS PBTA participants that do not have germline_sex_estimate:
`pbta_wxs_participants <- histologies_df %>% filter(cohort == "PBTA") %>% filter(experimental_strategy == "WXS") %>% filter(sample_type == "Tumor") %>% filter(is.na(germline_sex_estimate)) %>% select(-germline_sex_estimate)`
And then see whether those samples have WGS that have germline_sex_estimate that can be used:
`pbta_wgs_match <- histologies_df %>% filter(Kids_First_Participant_ID %in% pbta_wxs_participants$Kids_First_Participant_ID) %>% filter(experimental_strategy == "WGS") %>% select(Kids_First_Participant_ID, germline_sex_estimate)`
And merge the result back to the histology file.

But actually as I was going through this - it seemed like our histology file already have germline_sex_estimate updated to match WGS samples and the ones with NA are indeed samples without WGS.
@kgaonkar6, could you please check my logic and see whether that is the case for you as well?

Reviewer 3: Yes, the WXS which are paired WGS already has the germline_sex_estimate updated in the histology file. Sorry about the delay in my response!

Author: Removed the old erroneous lines of code, and implemented strategy as discussed below

  # mirroring controlfreec, X treated same as autosomes

  chrom == "chrX" & copy.num < tumor_ploidy & case_when ( cohort == "PBTA" & germline_sex_estimate == NA ~ samples_superset$germline_sex_estimate.x , TRUE ~ germline_sex_estimate) == "Female" ~ "loss",
  chrom == "chrX" & copy.num < tumor_ploidy & case_when (cohort == 'TARGET' & germline_sex_estimate == NA ~ case_when (samples_superset$germline_sex_estimate.x == NA ~ reported_gender, TRUE ~ samples_superset$germline_sex_estimate.x), TRUE ~ germline_sex_estimate) == "Female" ~ "loss",

  chrom == "chrX" & copy.num > tumor_ploidy & case_when ( cohort == "PBTA" & germline_sex_estimate == NA ~ samples_superset$germline_sex_estimate.x , TRUE ~ germline_sex_estimate) == "Female" ~ "gain",
  chrom == "chrX" & copy.num > tumor_ploidy & case_when (cohort == 'TARGET' & germline_sex_estimate == NA ~ case_when (samples_superset$germline_sex_estimate.x == NA ~ reported_gender, TRUE ~ samples_superset$germline_sex_estimate.x), TRUE ~ germline_sex_estimate) == "Female" ~ "gain",

>   chrom == "chrX" & copy.num == tumor_ploidy & case_when ( cohort == "PBTA" & germline_sex_estimate == NA ~ samples_superset$germline_sex_estimate.x , TRUE ~ germline_sex_estimate) == "Female" ~ "neutral"

Reviewer 3: I also think the easier way to handle this would be as @runjin326 suggested to first update the germline_sex_estimate to reported_gender where required ( where WXS germline_sex_estimate is NA) and then follow the original logic to update status. 

Here I see that a comma is also missing so that could be the source of the original error? Could you post the error here?

Author: Thank you @kgaonkar6 
The code works now and performs as expected.

    chrom == "chrX" & germline_sex_estimate == "Female" & copy.num == tumor_ploidy ~ "neutral"
  ))

> 
> 

Reviewer 2: ```suggestion
```

Author: Not implemented, change is now outdated.
